File: backend/routers/export.py
"""T-24 -- download this project's annotations in whichever format a
training pipeline actually wants. Reads straight out of PostgreSQL
(services/annotations_db.py); nothing here writes anything, and it doesn't
care whether the images came from the pool or the held-out test set.

Only three formats -- YOLO (what this tool always produced), COCO, and
Pascal VOC -- cover every consumer anyone's asked for so far (see
docs/DB_MIGRATION_PLAN.md #6). Add a fourth _export_x() the day someone
actually needs one; the dispatch below is a plain dict, not a registry."""
import io
import json
import zipfile
from pathlib import Path
from xml.sax.saxutils import escape

import cv2
from fastapi import APIRouter, HTTPException, Response

from ..deps import checked_path
from ..services import annotations_db

router = APIRouter(prefix="/api", tags=["export"])


def _dims(path: str) -> tuple[int, int] | None:
    """(width, height), or None if the file has moved/been deleted since it
    was annotated -- that image is skipped rather than failing the whole
    export."""
    img = cv2.imread(path)
    if img is None:
        return None
    h, w = img.shape[:2]
    return w, h


def _export_yolo(names: list[str], by_image: dict[str, list[dict]]) -> bytes:
    idx = {n: i for i, n in enumerate(names)}
    buf = io.BytesIO()
    with zipfile.ZipFile(buf, "w", zipfile.ZIP_DEFLATED) as zf:
        zf.writestr("classes.txt", "\n".join(names))
        for path, boxes in by_image.items():
            dims = _dims(path)
            if dims is None:
                continue
            w, h = dims
            lines = []
            for b in boxes:
                x1, y1, x2, y2 = b["box"]
                cx, cy = (x1 + x2) / 2 / w, (y1 + y2) / 2 / h
                bw, bh = abs(x2 - x1) / w, abs(y2 - y1) / h
                lines.append(f"{idx[b['cls']]} {cx:.6f} {cy:.6f} {bw:.6f} {bh:.6f}")
            zf.writestr(f"labels/{Path(path).stem}.txt", "\n".join(lines))
    return buf.getvalue()


def _export_coco(names: list[str], by_image: dict[str, list[dict]]) -> bytes:
    cat_id = {n: i + 1 for i, n in enumerate(names)}
    categories = [{"id": cid, "name": n} for n, cid in cat_id.items()]
    images, annotations = [], []
    ann_id = 1
    for image_id, (path, boxes) in enumerate(by_image.items(), start=1):
        dims = _dims(path)
        if dims is None:
            continue
        w, h = dims
        images.append({"id": image_id, "file_name": Path(path).name, "width": w, "height": h})
        for b in boxes:
            x1, y1, x2, y2 = b["box"]
            bw, bh = x2 - x1, y2 - y1
            annotations.append({
                "id": ann_id, "image_id": image_id, "category_id": cat_id[b["cls"]],
                "bbox": [x1, y1, bw, bh], "area": bw * bh, "iscrowd": 0,
            })
            ann_id += 1
    return json.dumps(
        {"images": images, "annotations": annotations, "categories": categories},
        ensure_ascii=False,
    ).encode("utf-8")


def _export_voc(names: list[str], by_image: dict[str, list[dict]]) -> bytes:
    buf = io.BytesIO()
    with zipfile.ZipFile(buf, "w", zipfile.ZIP_DEFLATED) as zf:
        for path, boxes in by_image.items():
            dims = _dims(path)
            if dims is None:
                continue
            w, h = dims
            objects = "".join(
                f"<object><name>{escape(b['cls'])}</name><bndbox>"
                f"<xmin>{b['box'][0]:.1f}</xmin><ymin>{b['box'][1]:.1f}</ymin>"
                f"<xmax>{b['box'][2]:.1f}</xmax><ymax>{b['box'][3]:.1f}</ymax>"
                "</bndbox></object>"
                for b in boxes
            )
            xml = (
                f"<annotation><filename>{escape(Path(path).name)}</filename>"
                f"<size><width>{w}</width><height>{h}</height><depth>3</depth></size>"
                f"{objects}</annotation>"
            )
            zf.writestr(f"{Path(path).stem}.xml", xml)
    return buf.getvalue()


_EXPORTERS = {
    "yolo": ("application/zip", "labels_yolo.zip", _export_yolo),
    "coco": ("application/json", "annotations_coco.json", _export_coco),
    "voc": ("application/zip", "labels_voc.zip", _export_voc),
}


@router.get("/export")
def export(input_dir: str, format: str = "yolo", kind: str = "pool"):
    if format not in _EXPORTERS:
        raise HTTPException(400, f"unknown format {format!r} -- choose one of {sorted(_EXPORTERS)}")
    if kind not in ("pool", "testset"):
        raise HTTPException(400, f"unknown kind {kind!r} -- choose 'pool' or 'testset'")
    inp = str(checked_path(input_dir))
    names = annotations_db.get_classes(inp, kind)
    by_image = annotations_db.load_annotations(inp, kind)
    if not names or not by_image:
        raise HTTPException(400, f"nothing to export for {kind!r} -- label something first")

    media, filename, build = _EXPORTERS[format]
    return Response(
        content=build(names, by_image), media_type=media,
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )

File: backend/internal/core/export/export.go
// Package export writes a project's annotations in whichever format a training
// pipeline actually wants.
//
// Ported from backend/the FastAPI export router (T-24). Pure read: nothing here writes
// state, and it does not care whether the images came from the pool or the
// held-out test set.
//
// Only three formats -- YOLO (what this tool always produced), COCO and Pascal
// VOC -- cover every consumer anyone has asked for. Add a fourth builder the day
// someone actually needs one; the dispatch is a plain map, not a registry.
//
// Coordinates come out of the database in pixels, so only YOLO and VOC reopen
// the image, and only for its dimensions. An image that has moved or been
// deleted since it was annotated is skipped rather than failing the whole
// export -- a stale row must not make the rest of a dataset unexportable.
package export

import (
	"archive/zip"
	"bytes"
	"encoding/json"
	"fmt"
	"path/filepath"
	"sort"
	"strings"

	"github.com/P-PrPas/CT-Flow/backend/internal/infra/store"
)

// Format is one supported output.
type Format struct {
	MediaType string
	Filename  string
	Build     func(names []string, byImage map[string][]store.Box, dims DimsFunc) ([]byte, error)
}

// DimsFunc reports an image's pixel dimensions, or false if it can no longer be
// read. Injected so the exporters stay testable without a filesystem.
type DimsFunc func(path string) (w, h int, ok bool)

var Formats = map[string]Format{
	"yolo": {"application/zip", "labels_yolo.zip", buildYOLO},
	"coco": {"application/json", "annotations_coco.json", buildCOCO},
	"voc":  {"application/zip", "labels_voc.zip", buildVOC},
}

// Names lists the supported formats, sorted, for the error message a bad
// request gets back.
func Names() []string {
	out := make([]string, 0, len(Formats))
	for k := range Formats {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

// sortedPaths fixes the iteration order. Python got this for free from a dict
// built by a query with ORDER BY i.path; a Go map has no order, and COCO image
// ids are assigned by position, so an unstable order would produce a different
// document every time.
func sortedPaths(byImage map[string][]store.Box) []string {
	paths := make([]string, 0, len(byImage))
	for p := range byImage {
		paths = append(paths, p)
	}
	sort.Strings(paths)
	return paths
}

func stem(path string) string {
	base := filepath.Base(path)
	return strings.TrimSuffix(base, filepath.Ext(base))
}

// f6 and f1 match Python's f"{v:.6f}" and f"{v:.1f}": fixed decimals, not the
// shortest representation.
func f6(v float64) string { return fmt.Sprintf("%.6f", v) }
func f1(v float64) string { return fmt.Sprintf("%.1f", v) }

func buildYOLO(names []string, byImage map[string][]store.Box, dims DimsFunc) ([]byte, error) {
	idx := make(map[string]int, len(names))
	for i, n := range names {
		idx[n] = i
	}

	var buf bytes.Buffer
	zw := zip.NewWriter(&buf)
	if err := writeZipEntry(zw, "classes.txt", strings.Join(names, "\n")); err != nil {
		return nil, err
	}
	for _, path := range sortedPaths(byImage) {
		w, h, ok := dims(path)
		if !ok {
			continue
		}
		lines := make([]string, 0, len(byImage[path]))
		for _, b := range byImage[path] {
			x1, y1, x2, y2 := b.Box[0], b.Box[1], b.Box[2], b.Box[3]
			cx, cy := (x1+x2)/2/float64(w), (y1+y2)/2/float64(h)
			bw, bh := abs(x2-x1)/float64(w), abs(y2-y1)/float64(h)
			lines = append(lines, fmt.Sprintf("%d %s %s %s %s",
				idx[b.Cls], f6(cx), f6(cy), f6(bw), f6(bh)))
		}
		if err := writeZipEntry(zw, "labels/"+stem(path)+".txt", strings.Join(lines, "\n")); err != nil {
			return nil, err
		}
	}
	if err := zw.Close(); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

type cocoImage struct {
	ID       int    `json:"id"`
	FileName string `json:"file_name"`
	Width    int    `json:"width"`
	Height   int    `json:"height"`
}

type cocoAnnotation struct {
	ID         int        `json:"id"`
	ImageID    int        `json:"image_id"`
	CategoryID int        `json:"category_id"`
	BBox       [4]float64 `json:"bbox"`
	Area       float64    `json:"area"`
	IsCrowd    int        `json:"iscrowd"`
}

type cocoCategory struct {
	ID   int    `json:"id"`
	Name string `json:"name"`
}

func buildCOCO(names []string, byImage map[string][]store.Box, dims DimsFunc) ([]byte, error) {
	catID := make(map[string]int, len(names))
	categories := make([]cocoCategory, 0, len(names))
	for i, n := range names {
		catID[n] = i + 1 // COCO category ids are 1-based
		categories = append(categories, cocoCategory{ID: i + 1, Name: n})
	}

	imgs := []cocoImage{}
	anns := []cocoAnnotation{}
	annID := 1
	for i, path := range sortedPaths(byImage) {
		// The id counts positions, not emitted images: a skipped image consumes
		// its number and leaves a gap. That is what the Python enumerate() did,
		// and COCO only requires ids to be unique, not contiguous.
		imageID := i + 1
		w, h, ok := dims(path)
		if !ok {
			continue
		}
		imgs = append(imgs, cocoImage{ID: imageID, FileName: filepath.Base(path), Width: w, Height: h})
		for _, b := range byImage[path] {
			bw, bh := b.Box[2]-b.Box[0], b.Box[3]-b.Box[1]
			anns = append(anns, cocoAnnotation{
				ID: annID, ImageID: imageID, CategoryID: catID[b.Cls],
				BBox: [4]float64{b.Box[0], b.Box[1], bw, bh}, Area: bw * bh, IsCrowd: 0,
			})
			annID++
		}
	}

	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	// Python wrote this with ensure_ascii=False; Go escapes <, > and & in
	// strings unless told not to, which would mangle a class name containing
	// one for no reason.
	enc.SetEscapeHTML(false)
	if err := enc.Encode(map[string]any{
		"images": imgs, "annotations": anns, "categories": categories,
	}); err != nil {
		return nil, err
	}
	return bytes.TrimRight(buf.Bytes(), "\n"), nil
}

func buildVOC(names []string, byImage map[string][]store.Box, dims DimsFunc) ([]byte, error) {
	var buf bytes.Buffer
	zw := zip.NewWriter(&buf)
	for _, path := range sortedPaths(byImage) {
		w, h, ok := dims(path)
		if !ok {
			continue
		}
		var objects strings.Builder
		for _, b := range byImage[path] {
			objects.WriteString("<object><name>" + xmlEscape(b.Cls) + "</name><bndbox>" +
				"<xmin>" + f1(b.Box[0]) + "</xmin><ymin>" + f1(b.Box[1]) + "</ymin>" +
				"<xmax>" + f1(b.Box[2]) + "</xmax><ymax>" + f1(b.Box[3]) + "</ymax>" +
				"</bndbox></object>")
		}
		doc := "<annotation><filename>" + xmlEscape(filepath.Base(path)) + "</filename>" +
			fmt.Sprintf("<size><width>%d</width><height>%d</height><depth>3</depth></size>", w, h) +
			objects.String() + "</annotation>"
		if err := writeZipEntry(zw, stem(path)+".xml", doc); err != nil {
			return nil, err
		}
	}
	if err := zw.Close(); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

// xmlEscape matches Python's xml.sax.saxutils.escape: &, < and > only.
// encoding/xml's escaper also rewrites quotes and newlines, which would make
// these documents differ from every one this tool has produced so far.
// Ampersand first, or the entities introduced by the other two get re-escaped.
func xmlEscape(s string) string {
	s = strings.ReplaceAll(s, "&", "&amp;")
	s = strings.ReplaceAll(s, "<", "&lt;")
	return strings.ReplaceAll(s, ">", "&gt;")
}

func writeZipEntry(zw *zip.Writer, name, body string) error {
	// Deflate, matching zipfile.ZIP_DEFLATED -- the default here is Store.
	w, err := zw.CreateHeader(&zip.FileHeader{Name: name, Method: zip.Deflate})
	if err != nil {
		return err
	}
	_, err = w.Write([]byte(body))
	return err
}

func abs(v float64) float64 {
	if v < 0 {
		return -v
	}
	return v
}
